src/libtls/tls_hkdf.c

```diff
@@ -0,0 +1,602 @@
+{
+	/*
+	 * pack label and context into structure hkdflabel
+	 */
+	if (label->len < 7 || label->len > 255 ||
+		context->len < 0 || context->len > 255)
+	{
+		return FALSE;
+	}
+	/* HKDFLabel */
+	size_t offset = 0;
+	size_t chunk_len = 2 + 1 + label->len + 1 + context->len;
+	this->info = chunk_create(malloc(chunk_len), chunk_len);
+	this->info.ptr[1] = length & 0xFF;
+	this->info.ptr[0] = length >> 8;
+	this->info.ptr[2] = label->len;
+	offset += 3;
+	memcpy(&(this->info).ptr[offset], label->ptr, label->len);
+	offset += label->len;
+	this->info.ptr[offset] = context->len;
+	offset += 1;
+	memcpy(&(this->info).ptr[offset], context->ptr, context->len);
+	offset += context->len;
+
+	expand(this);
+
+	return TRUE;
+}
+
+static bool derive_secret(private_tls_hkdf_t *this, chunk_t *label, chunk_t *messages)
+{
+	/*
+	 * calculate hash of message
+	 */
+	/*
+	 * Hash messages and call expand
+	 */
+	size_t chunk_len = 32;
+	chunk_t context = chunk_create(malloc(chunk_len), chunk_len);
+
+	if (!this->hasher || !this->hasher->allocate_hash(this->hasher, *messages, &context))
+	{
+		DBG2(DBG_TLS, "%N not supported", hash_algorithm_names, this->hash_algorithm);
+		return FALSE;
+	}
+
+	expand_label(this, label, &context, 32);
+	chunk_free(&context);
+
+	return TRUE;
+}
+
+static bool move_to_phase_1(private_tls_hkdf_t *this)
+{
+	if (this->phase == HKDF_PHASE_0)
+	{
+		if (!extract(this))
+		{
+			DBG2(DBG_TLS, "unable extract PRK");
+			return FALSE;
+		}
+		this->phase = HKDF_PHASE_1;
+		/* prepare okm for next extract */
+		chunk_t empty = chunk_empty;
+		chunk_t derived = chunk_from_str("tls13 derived");
+		if (!derive_secret(this, &derived, &empty))
+		{
+			DBG2(DBG_TLS, "unable to derive secret");
+			return FALSE;
+		}
+		return TRUE;
+	}
+
+	if (this->phase == HKDF_PHASE_1)
+	{
+		return TRUE;
+	}
+	else
+	{
+		DBG2(DBG_TLS, "invalid HKDF phase");
+		return FALSE;
+	}
+}
+
+static bool move_to_phase_2(private_tls_hkdf_t *this)
+{
+	if (this->phase == HKDF_PHASE_0)
+	{
+		move_to_phase_1(this);
+	}
+
+	if (this->phase == HKDF_PHASE_1)
+	{
+		this->salt = this->okm;
+
+		if (!this->shared_secret)
+		{
+			DBG2(DBG_TLS, "no shared secret set");
+			return FALSE;
+		}
+		else
+		{
+			memcpy(this->ikm.ptr, this->shared_secret->ptr, this->shared_secret->len);
+			this->ikm.len = this->shared_secret->len;
+		}
+
+		if (!extract(this))
+		{
+			DBG2(DBG_TLS, "unable extract PRK");
+			return FALSE;
+		}
+		this->phase = HKDF_PHASE_2;
+		/* prepare okm for next extract */
+		chunk_t empty = chunk_empty;
+		chunk_t derived = chunk_from_str("tls13 derived");
+		if (!derive_secret(this, &derived, &empty))
+		{
+			DBG2(DBG_TLS, "unable to derive secret");
+			return FALSE;
+		}
+		return TRUE;
+	}
+
+	if (this->phase == HKDF_PHASE_2)
+	{
+		return TRUE;
+	}
+	else
+	{
+		DBG2(DBG_TLS, "invalid HKDF phase");
+		return FALSE;
+	}
+}
+
+/*
+static bool move_to_phase_3(private_tls_hkdf_t *this)
```

Comment:
Maybe add an explanation why this is commented out - it's not in scope/we haven't arrived at this part yet.